packages/events/src/decorators.ts

```diff
@@ -0,0 +1,286 @@
+import {
+  navigateType,
+  type Diagnostic,
+  type Model,
+  type ModelProperty,
+  type Program,
+  type Tuple,
+  type Type,
+  type Union,
+  type UnionVariant,
+} from "@typespec/compiler";
+import type {
+  ContentTypeDecorator,
+  DataDecorator,
+  EventsDecorator,
+} from "../generated-defs/TypeSpec.Events.js";
+import { createDiagnostic, EventsStateKeys } from "./lib.js";
+
+/** @internal */
+export const namespace = "TypeSpec.Events";
+
+export const $events: EventsDecorator = (context, target) => {
+  context.program.stateSet(EventsStateKeys.events).add(target);
+};
+
+export function isEvents(program: Program, target: Union): boolean {
+  return program.stateSet(EventsStateKeys.events).has(target);
+}
+
+export const $contentType: ContentTypeDecorator = (context, target, contentType) => {
+  context.program.stateMap(EventsStateKeys.contentType).set(target, contentType);
+};
+
+export function getContentType(
+  program: Program,
+  target: UnionVariant | ModelProperty,
+): string | undefined {
+  return program.stateMap(EventsStateKeys.contentType).get(target);
+}
+
+function validateContentType(program: Program, target: ModelProperty): Diagnostic[] {
+  const contentType = getContentType(program, target);
+  if (!contentType || isEventData(program, target)) return [];
+
+  return [
+    createDiagnostic({
+      code: "invalid-content-type-target",
+      target,
+    }),
+  ];
+}
+
+export const $data: DataDecorator = (context, target) => {
+  context.program.stateSet(EventsStateKeys.data).add(target);
+};
+
+export function isEventData(program: Program, target: ModelProperty): boolean {
+  return program.stateSet(EventsStateKeys.data).has(target);
+}
+
+function stringifyPropertyPath(path: Array<Model | ModelProperty | Tuple>): string {
+  return path
+    .map((p) => {
+      switch (p.kind) {
+        case "ModelProperty":
+          return p.name;
+        case "Tuple":
+          return "[]";
+        case "Model":
+          return;
+      }
+    })
+    .filter(Boolean)
+    .join(".");
+}
+
+function getEventDefinition(
+  program: Program,
+  target: UnionVariant,
+): [EventDefinition, readonly Diagnostic[]] {
+  const diagnostics: Diagnostic[] = [];
+
+  // If the variant has a name, that may be used for out-of-band event types.
+  const eventType = typeof target.name === "string" ? target.name : undefined;
+  const envelopeContentType = getContentType(program, target);
+
+  let payloadType: Type | undefined;
+  let payloadContentType: string | undefined;
+  let pathToPayload: string = "";
```

Comment:
Yes, just for diagnostics for now. I was debating if it would be useful for emitters too - though I figured they'd need to crawl the event envelope anyway so could compare the payloadType to whatever node they're in.